cpp/src/strings/utilities.cu

```diff
@@ -145,20 +145,42 @@ std::unique_ptr<column> make_empty_strings_column( rmm::mr::device_memory_resour
 namespace
 {
 
-std::mutex g_flags_table_mutex;
-static character_flags_table_type* d_character_flags_table = nullptr;
+// guaranteed thread-safe
+struct character_flags_singleton
+{
+    static character_flags_singleton& instance()
+    {
+        static character_flags_singleton _instance;
+        return _instance;
+    }
+    character_flags_singleton( character_flags_singleton const& ) = delete;
+    character_flags_singleton( character_flags_singleton&& ) = delete;
+    character_flags_singleton& operator=(character_flags_singleton const&) = delete;
+    character_flags_singleton& operator=(character_flags_singleton&&) = delete;
 
-} // namespace
+    character_flags_table_type* table() { return d_character_flags_table; }
 
-const character_flags_table_type* get_character_flags_table()
-{
-    std::lock_guard<std::mutex> guard(g_flags_table_mutex);
-    if( !d_character_flags_table )
+private:
+
+    character_flags_table_type* d_character_flags_table{};
+
+    character_flags_singleton()
     {
         RMM_TRY(RMM_ALLOC(&d_character_flags_table,sizeof(g_character_codepoint_flags),0));
         CUDA_TRY(cudaMemcpy(d_character_flags_table,g_character_codepoint_flags,sizeof(g_character_codepoint_flags),cudaMemcpyHostToDevice));
     }
-    return d_character_flags_table;
+    ~character_flags_singleton()
+    {
+        // will this be guaranteed to be called before rmmFinalize?
```

Comment:
Great question. It appears that the answer is yes: https://stackoverflow.com/a/16010180/11341974

Reply:
Or another solution is just to use `cudaMalloc/cudaFree` instead of RMM.

Reply:
Yes, I did that in NVStrings to keep this memory from being added to the RMM pool. It's only about 200KB. I worried about the case where someone gives all the GPU memory to the RMM pool and then these would fail.

Reply:
I wonder if we could just not worry about it...  Currently `rmm::free()` returns `RMM_NOT_INITIALIZED` in this situation and does not attempt to free the pointer. In the near future when we switch to exceptions, `rmm::free()` will *not* throw, since it's not allowed to throw in a destructor and deallocation is common in destructors. Therefore, it will become a no-op.

In pool mode, there will be no memory leak, because finalizing the pool will free it.

In non-pool mode, there will be a memory leak, because it will not call `cudaFree()` in this case...  So here's a possible workaround. Check if RMM is in pool mode, and use RMM_ALLOC if it is.  If not, call cudaMalloc directly.

That way you don't risk failing to allocate when RMM reserves all GPU memory, and you don't risk failing to deallocate when there's no pool.

Not beautiful, but it works...

Reply:
Or perhaps we need to make `rmm::free()` just a pass-through when not in pool mode.